src/mod/asr_tts/mod_unimrcp/mod_unimrcp.c

```diff
@@ -1926,7 +1926,7 @@ static apt_bool_t speech_on_channel_add(mrcp_application_t *application, mrcp_se
 	} else {
 		descriptor = mrcp_application_source_descriptor_get(channel);
 	}
-	if (!descriptor) {
+	if (!descriptor || !descriptor->sampling_rate || schannel->channel_destroyed) {
```

Comment:
Why do we need to evaluate sampling_rate here?

Reply:
I added it as an extra precaution to check if a null parameter is passed as `sampling_rate` while the session is still on. The stack track indicated a possible null pointer passed to `switch_log_printf` which caused a seg fault. And since we checked `descriptor` and `schannel` already, I think the possible issue could be the returned data from the mrcp applications

```
Stack trace:
0 [__strlen_evex()] sysdeps/x86_64/multiarch/strlen-evex.S:77
1 [__vfprintf_internal()] ./stdio-common/vfprintf-internal.c:1647 (discriminator 57)
2 [__vasprintf_internal()] ./libio/vasprintf.c:57
3 [switch_vasprintf()] /SRC/freeswitch/src/switch_apr.c:1299
4 [switch_log_meta_vprintf()] /SRC/freeswitch/src/switch_log.c:645
5 [switch_log_printf()] /SRC/freeswitch/src/switch_log.c:554
6 [speech_on_channel_add()] /SRC/freeswitch/src/mod/asr_tts/mod_unimrcp/mod_unimrcp.c:1945
7 [mrcp_application_message_dispatch()] /usr/local/freeswitch/lib/freeswitch/mod/mod_unimrcp.so:0x2f234
8 [recog_message_handler()] /SRC/freeswitch/src/mod/asr_tts/mod_unimrcp/mod_unimrcp.c:3659
9 [mrcp_app_sig_response_raise()] mrcp_client_session.c:?
10 [mrcp_client_mpf_message_process()] /usr/local/freeswitch/lib/freeswitch/mod/mod_unimrcp.so:0x34628
11 [mrcp_client_msg_process()] mrcp_client.c:?
12 [apt_task_msg_process()] /usr/local/freeswitch/lib/freeswitch/mod/mod_unimrcp.so:0x40f48
13 [apt_consumer_task_run()] apt_consumer_task.c:?
14 [apt_task_run()] apt_task.c:?
15 [dummy_worker()] thread.c:?
```

Reply:
sampling_rate is an interger, it would not cause the crash, the name may have been the cause of the crash since its a pointer. The issue might be is this command
```
speech_channel_t *schannel = (speech_channel_t *) mrcp_application_channel_object_get(channel);
``` 

If somehow FS already destroy schannel but since we got a timeout on terminating mrcp then this mrcp may still hold a garbage pointer. Since its a garbage pointer then the validation to check of schannel and channel_destroyed may return incorrect
```
	if (!session || !schannel || status != MRCP_SIG_STATUS_CODE_SUCCESS || schannel->channel_destroyed) {
		goto error;
	}
```

What I think which will be a best solution is when we encounter a timeout for creating and destroying mrcp session, we should set the schannel or the private obj to null in the mrcp session. We should create a new method mrcp_application_channel_object_set which we can pass a NULL value. Just for reference this is the mrcp_application_channel_object_get.
```
MRCP_DECLARE(void*) mrcp_application_channel_object_get(const mrcp_channel_t *channel)
{
	if(!channel) {
		return FALSE;
	}
	return channel->obj;
}
```

Reply:
The channel obj and schannel are from the mrcp `app_message` which is received from mrcp responses. Setting the mrcp channel to NULL in cases of timeouts might not work here as the channel data will be overriden when responses are received from mrcp client.
I implemented another alternative to check if freeswitch session is  still active when the mrcp data is processed

Reply:
if schannel is already a garbage pointer (not null value) then session_uuid will also be a garbage value. The crash may happened in the switch_core_session_locate.

Reply:
But won't the session_uuid just be a key that doesn't exist on the hash table, which then returns NULL?

Reply:
schannel and session_uuid lifetime is based on the switch_speech_handle_t pool. So if the pool has been deleted then both schannel and session_uuid is already a garbage pointer. If you access a garbage pointer, it would result to a crash